File: web/src/types/datahub.ts
export interface DatahubDatasetInfo {
  id: string
  dataset_key: string
  label_zh?: string | null
  layer: string
  schema_version: string
  description?: string | null
  is_active: string
  updated_at: string
}

export interface DatahubJobRunInfo {
  id: string
  job_type: string
  dataset?: string | null
  status: string
  trigger_source?: string | null
  started_at?: string | null
  finished_at?: string | null
  task_total: number
  task_success: number
  task_failed: number
  error_message?: string | null
  created_at: string
  updated_at: string
}

export interface DatahubJobTaskInfo {
  id: string
  job_run_id: string
  dataset: string
  symbol?: string | null
  start_date?: string | null
  end_date?: string | null
  status: string
  attempts: number
  last_error?: string | null
  locked_at?: string | null
  created_at: string
  updated_at: string
}

export interface DatahubQualityReportInfo {
  id: string
  dataset: string
  symbol?: string | null
  biz_date?: string | null
  quality_score: number
  severity: string
  issues?: Array<Record<string, unknown>> | null
  object_key?: string | null
  checked_at: string
  created_at: string
}

export interface DatahubObjectIndexInfo {
  id: string
  bucket: string
  object_key: string
  dataset: string
  layer: string
  provider?: string | null
  symbol?: string | null
  start_date?: string | null
  end_date?: string | null
  row_count: number
  schema_version: string
  content_hash?: string | null
  quality_score?: number | null
  created_at: string
}

export interface DatahubWorkerHeartbeatInfo {
  worker_name: string
  status: string
  last_heartbeat_at: string
  last_run_id?: string | null
  processed_count: number
  last_error?: string | null
  is_online: boolean
  offline_threshold_seconds: number
}

export interface DatahubProviderHealthInfo {
  provider: string
  dataset: string
  status: string
  success_count: number
  failure_count: number
  failure_rate: number
  last_success_at?: string | null
  last_failure_at?: string | null
  last_error?: string | null
  cooldown_until?: string | null
  is_available: boolean
}

export interface DatahubMetricsSummaryInfo {
  window_days: number
  total_runs: number
  success_runs: number
  failed_runs: number
  running_runs: number
  success_rate: number
  avg_duration_seconds: number
  p95_duration_seconds: number
  avg_quality_score: number
  p0_quality_count: number
  provider_cooldown_count: number
  provider_degraded_count: number
}

export interface TriggerBackfillPayload {
  dataset: string
  start_date: string
  end_date: string
  symbol?: string
  symbols?: string[]
}

export interface TriggerDailyIncrementalPayload {
  dataset: string
  symbol?: string
  window_days: number
}

export interface PurgeJobRunsPayload {
  status: string
  limit?: number
}

export interface PurgeJobRunsResult {
  deleted_count: number
}

export interface DatahubFailureGroupInfo {
  error_message: string
  count: number
  symbols: string[]
}

export interface DatahubRunFailureDetailInfo {
  run_id: string
  failed_count: number
  groups: DatahubFailureGroupInfo[]
  tasks: DatahubJobTaskInfo[]
}

export interface RetryFailedTasksPayload {
  strategy: 'immediate' | 'by_error'
  max_retry_attempts: number
}

export interface RetryFailedTasksResult {
  created_runs: number
  skipped_tasks: number
  retried_tasks: number
  retried_symbols: string[]
}

export interface MarketDailyMissingScanResult {
  dataset: string
  start_date: string
  end_date: string
  reference_date: string
  expected_count: number
  existing_count: number
  missing_count: number
  missing_symbols: string[]
}

export interface FillMarketDailyMissingPayload {
  start_date: string
  end_date: string
  reference_date?: string
  max_symbols: number
  batch_size: number
}

export interface FillMarketDailyMissingResult {
  created_runs: number
  filled_symbols: number
  symbols: string[]
}

export interface DatahubWatchlistInfo {
  id: string
  symbol: string
  name?: string | null
  sort_order: number
  note?: string | null
  backfill_start_date?: string | null
  backfill_end_date?: string | null
  window_days?: number | null
  created_at: string
  updated_at: string
}

export interface DatahubWatchlistUpdatePayload {
  name?: string
  note?: string
}

export interface MarketDailyBarInfo {
  symbol: string
  trade_date: string
  open: number
  high: number
  low: number
  close: number
  volume: number
  amount: number
  turnover_rate?: number | null
}

export interface WatchlistBoardRow {
  id: string
  symbol: string
  name?: string | null
  sector_name?: string | null
  trade_date?: string | null
  open?: number | null
  high?: number | null
  low?: number | null
  close?: number | null
  change_amount?: number | null
  change_pct?: number | null
  sector_change_pct?: number | null
  volume?: number | null
  turnover_rate?: number | null
  has_data: boolean
}

export interface WatchlistBoardResponse {
  limit_days: number
  rows: WatchlistBoardRow[]
}

export interface DatahubWatchlistWatermarkInfo {
  dataset: string
  dataset_label: string
  last_success_date?: string | null
  last_quality_score?: number | null
}

export interface DatahubWatchlistSymbolSummary {
  symbol: string
  name?: string | null
  backfill_start_date?: string | null
  backfill_end_date?: string | null
  window_days?: number | null
  market_daily_start_date?: string | null
  market_daily_end_date?: string | null
  market_daily_row_count: number
  market_daily_quality_score?: number | null
  latest_quality_score?: number | null
  latest_quality_severity?: string | null
  watermarks: DatahubWatchlistWatermarkInfo[]
  object_indexes: Array<{
    dataset: string
    dataset_label: string
    start_date?: string | null
    end_date?: string | null
    row_count: number
    quality_score?: number | null
    object_key: string
  }>
  quality_reports: Array<{
    dataset: string
    dataset_label: string
    biz_date?: string | null
    quality_score: number
    severity: string
    checked_at?: string | null
  }>
}

export interface DailyBriefReadinessInfo {
  as_of_date: string
  status: string
  latest_trade_date?: string | null
  watchlist_count: number
  missing_count: number
  worker_online: boolean
  warnings: string[]
}

export interface DailyBriefMarketInfo {
  indices: Array<Record<string, unknown>>
  five_day_trend: string
  turnover_summary?: Record<string, unknown> | null
  breadth?: Record<string, unknown> | null
}

export interface DailyBriefSectorInfo {
  sector_name: string
  stock_count: number
  avg_stock_change_pct?: number | null
  sector_change_pct?: number | null
}

export interface DailyBriefWatchlistItemInfo {
  symbol: string
  name?: string | null
  sector_name?: string | null
  trade_date?: string | null
  close?: number | null
  change_pct?: number | null
  sector_change_pct?: number | null
  volume?: number | null
  turnover_rate?: number | null
  has_data: boolean
  data_status: string
}

export interface DailyBriefRiskFlagInfo {
  level: string
  source: string
  message: string
  symbol?: string | null
}

export interface DailyBriefContextInfo {
  as_of_date: string
  readiness: DailyBriefReadinessInfo
  market: DailyBriefMarketInfo
  related_sectors: DailyBriefSectorInfo[]
  watchlist: DailyBriefWatchlistItemInfo[]
  portfolio?: Record<string, unknown> | null
  observations: Array<Record<string, unknown>>
  risk_flags: DailyBriefRiskFlagInfo[]
  data_quality: Record<string, string>
}

export interface DailyBriefPrepareResultInfo {
  created_runs: number
  runs: DatahubJobRunInfo[]
  message: string
}

File: api/app/datahub/schemas/datahub.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field, model_validator


class DatahubDatasetInfo(BaseModel):
    id: str
    dataset_key: str
    label_zh: Optional[str] = None
    layer: str
    schema_version: str
    description: Optional[str] = None
    is_active: str
    updated_at: datetime


class DatahubJobRunInfo(BaseModel):
    id: str
    job_type: str
    dataset: Optional[str] = None
    status: str
    trigger_source: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    task_total: int
    task_success: int
    task_failed: int
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class PurgeJobRunsRequest(BaseModel):
    status: str = Field(..., description="按该状态批量删除运行记录，如 failed")
    limit: int = Field(100, ge=1, le=500, description="最多删除条数")


class PurgeJobRunsResult(BaseModel):
    deleted_count: int


class DatahubFailureGroupInfo(BaseModel):
    error_message: str
    count: int
    symbols: list[str] = Field(default_factory=list)


class DatahubRunFailureDetailInfo(BaseModel):
    run_id: str
    failed_count: int
    groups: list[DatahubFailureGroupInfo]
    tasks: list["DatahubJobTaskInfo"]


class RetryFailedTasksRequest(BaseModel):
    strategy: str = Field("immediate", description="重试策略：immediate/by_error")
    max_retry_attempts: int = Field(3, ge=1, le=10, description="超过该重试次数的失败任务将跳过")


class RetryFailedTasksResult(BaseModel):
    created_runs: int
    skipped_tasks: int
    retried_tasks: int
    retried_symbols: list[str] = Field(default_factory=list)


class MarketDailyMissingScanResult(BaseModel):
    dataset: str = "market_daily"
    start_date: date
    end_date: date
    reference_date: date
    expected_count: int
    existing_count: int
    missing_count: int
    missing_symbols: list[str] = Field(default_factory=list)


class FillMarketDailyMissingRequest(BaseModel):
    start_date: date
    end_date: date
    reference_date: Optional[date] = None
    max_symbols: int = Field(500, ge=1, le=5000)
    batch_size: int = Field(200, ge=1, le=1000)


class FillMarketDailyMissingResult(BaseModel):
    created_runs: int
    filled_symbols: int
    symbols: list[str] = Field(default_factory=list)


class DatahubJobTaskInfo(BaseModel):
    id: str
    job_run_id: str
    dataset: str
    symbol: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: str
    attempts: int
    last_error: Optional[str] = None
    locked_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class DatahubQualityReportInfo(BaseModel):
    id: str
    dataset: str
    symbol: Optional[str] = None
    biz_date: Optional[date] = None
    quality_score: float
    severity: str
    issues: Optional[list[dict]] = None
    object_key: Optional[str] = None
    checked_at: datetime
    created_at: datetime


class DatahubObjectIndexInfo(BaseModel):
    id: str
    bucket: str
    object_key: str
    dataset: str
    layer: str
    provider: Optional[str] = None
    symbol: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    row_count: int
    schema_version: str
    content_hash: Optional[str] = None
    quality_score: Optional[float] = None
    created_at: datetime


class DatahubWorkerHeartbeatInfo(BaseModel):
    worker_name: str
    status: str
    last_heartbeat_at: datetime
    last_run_id: Optional[str] = None
    processed_count: int
    last_error: Optional[str] = None
    is_online: bool
    offline_threshold_seconds: int


class DatahubProviderHealthInfo(BaseModel):
    provider: str
    dataset: str
    status: str
    success_count: int
    failure_count: int
    failure_rate: float
    last_success_at: Optional[datetime] = None
    last_failure_at: Optional[datetime] = None
    last_error: Optional[str] = None
    cooldown_until: Optional[datetime] = None
    is_available: bool


class DatahubMetricsSummaryInfo(BaseModel):
    window_days: int
    total_runs: int
    success_runs: int
    failed_runs: int
    running_runs: int
    success_rate: float
    avg_duration_seconds: float
    p95_duration_seconds: float
    avg_quality_score: float
    p0_quality_count: int
    provider_cooldown_count: int
    provider_degraded_count: int


class TriggerBackfillRequest(BaseModel):
    dataset: str = Field(..., description="数据集")
    start_date: date = Field(..., description="开始日期")
    end_date: date = Field(..., description="结束日期")
    symbol: Optional[str] = Field(None, description="证券代码，可选")
    symbols: Optional[list[str]] = Field(None, description="证券代码列表，可选，优先于 symbol")

    @model_validator(mode="after")
    def validate_symbol_fields(self) -> "TriggerBackfillRequest":
        if self.symbol and self.symbols:
            raise ValueError("symbol 与 symbols 不能同时传入")
        return self


class TriggerDailyIncrementalRequest(BaseModel):
    dataset: str = Field("market_daily", description="数据集")
    symbol: Optional[str] = Field(None, description="证券代码，可选")
    window_days: int = Field(7, ge=1, le=30, description="回溯窗口天数")


class DatahubWatchlistCreate(BaseModel):
    symbol: str = Field(..., description="证券代码，支持 6 位数字或带后缀格式")
    name: Optional[str] = Field(None, description="证券名称，可选")
    note: Optional[str] = Field(None, description="备注，可选")


class DatahubWatchlistInfo(BaseModel):
    id: str
    symbol: str
    name: Optional[str] = None
    sort_order: int
    note: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class DatahubWatchlistUpdate(BaseModel):
    name: Optional[str] = Field(None, description="证券名称")
    note: Optional[str] = Field(None, description="备注")


class DatahubWatchlistWatermarkInfo(BaseModel):
    dataset: str
    dataset_label: str
    last_success_date: Optional[date] = None
    last_quality_score: Optional[float] = None


class DatahubWatchlistSymbolSummary(BaseModel):
    symbol: str
    name: Optional[str] = None
    market_daily_start_date: Optional[date] = None
    market_daily_end_date: Optional[date] = None
    market_daily_row_count: int = 0
    market_daily_quality_score: Optional[float] = None
    latest_quality_score: Optional[float] = None
    latest_quality_severity: Optional[str] = None
    watermarks: list[DatahubWatchlistWatermarkInfo] = Field(default_factory=list)
    object_indexes: list[dict] = Field(default_factory=list)
    quality_reports: list[dict] = Field(default_factory=list)


class MarketDailyBarInfo(BaseModel):
    symbol: str
    trade_date: date
    open: float
    high: float
    low: float
    close: float
    volume: float
    amount: float
    turnover_rate: Optional[float] = None


class WatchlistBoardRow(BaseModel):
    id: str
    symbol: str
    name: Optional[str] = None
    sector_name: Optional[str] = None
    trade_date: Optional[date] = None
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None
    change_amount: Optional[float] = None
    change_pct: Optional[float] = None
    sector_change_pct: Optional[float] = None
    volume: Optional[float] = None
    turnover_rate: Optional[float] = None
    has_data: bool = False


class WatchlistBoardResponse(BaseModel):
    limit_days: int
    rows: list[WatchlistBoardRow] = Field(default_factory=list)
